### PR TITLE
Pub 985 - Summary of publications / lists

### PR DESCRIPTION
**Pull request by furnivall**
### JIRA links ###
https://tools.hmcts.net/jira/browse/PUB-985
https://tools.hmcts.net/jira/browse/PUB-983
https://tools.hmcts.net/jira/browse/PUB-922
https://tools.hmcts.net/jira/browse/PUB-944

### Change description ###

- Implemented summary of publications for single justice procedure and all courts from alphabetical-search-page.

- Implemented placeholder logic for case of only 1 publication being returned (from 922).

- Implemented logic to differentiate between PDFs and JSON files (from 983)

SJP page:

image

Aberdeen Court page:

image

### Notes: ### 
- Have kept the old SJP controller and List-option controllers so they can be worked on for 984 and 997.
- SingleJusticeProcedure route has been removed but can be re-added without too much issue if desired.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ x] No
```

### CODE REVIEW

File: src/main/resources/requests/publicationRequests.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
+import fs from 'fs';
+import path from 'path';
+import {dataManagementApi} from './utils/axiosConfig';
+import {Publication} from '../../models/publication';
+
+export class PublicationRequests {
+  mocksPath = '../mocks';
+  rawData = fs.readFileSync(path.resolve(__dirname, this.mocksPath, 'SingleJusticeProcedureCases.json'), 'utf-8');
```
Comment (ChrisS1512) [inline src/main/resources/requests/publicationRequests.ts:8]: Lines 7 and 8 can go now :)

**Review by ChrisS1512 — CHANGES_REQUESTED** [review]

File: src/test/unit/controllers/PublicationController.test.ts [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    };
+
+    responseMock.expects('render').once().withArgs('summary-of-publications', expectedData);
+
+    await publicationController.get(request, response);
+    responseMock.verify();
+  });
+});
+
+describe('Get publications2', () => {
```
Comment (ChrisS1512) [inline src/test/unit/controllers/PublicationController.test.ts:78]: I'd rename this

**Review by ChrisS1512 — CHANGES_REQUESTED** [review]

File: src/main/resources/locales/en/account-home.json [hunk 3]
```diff
@@ -15,7 +15,7 @@
     {
       "title": "Single Justice Procedure cases",
       "description": "A nationwide list of cases, ready to be decided by a magistrate without a hearing. Such as TV licensing prosecutions.",
-      "link": "single-justice-procedure"
+      "link": "publication?courtId=0"
```
Comment (kurt-uk) [inline src/main/resources/locales/en/account-home.json:18]: the link can live in the nunjuks file as this locales is for when the language switches from english to welsh, and the link will be the same regardless of the language

File: src/main/views/old-single-justice-procedure.njk [hunk 4]
```diff
@@ -0,0 +1,79 @@
+{% from "./macros/common-components.njk" import goBack %}
```
Comment (kurt-uk) [inline src/main/views/old-single-justice-procedure.njk:1]: bin off?

File: src/main/views/summary-of-publications.njk [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+  {% for item in publications |sort(reverse= true, attribute="contentDate, listType")%}
+
+  <a class = das-search-results__link href = {{item.artefactId}}>{{ item.listType | listType }}
+  {{item.contentDate | date("DD MMMM YYYY")}}</a>
+  {% if item.isFlatFile %}
+  (opens in a new window)
+  {% endif %}
+
+  <br>
```
Comment (kurt-uk) [inline src/main/views/summary-of-publications.njk:32]: can we extend with the margin classes at all or does it have to be breaks?

File: src/test/unit/controllers/PublicationController.test.ts [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+const rawSJPData = fs.readFileSync(path.resolve(__dirname, '../mocks/trimmedSJPCases.json'), 'utf-8');
+const sjpCases = JSON.parse(rawSJPData).results;
+const onePubData = fs.readFileSync(path.resolve(__dirname, '../mocks/onePublication.json'), 'utf-8');
+const onePub = JSON.parse(onePubData);
+describe('Get publications', () => {
+  beforeEach(function() {
+    sinon.stub(PublicationService.prototype, 'getPublications').resolves(sjpCases);
+  });
+  afterEach(function() {
+    sinon.restore();
```
Comment (kurt-uk) [inline src/test/unit/controllers/PublicationController.test.ts:22]: how come you keep defining and restoring each test? it doesnt look like you change the value at all so could just move line 19 above the describe and have it class level

File: src/test/unit/controllers/PublicationController.test.ts [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+  });
+
+  it('should render the Summary of Publications page', async () => {
+
+    const response = {
```
Comment (kurt-uk) [inline src/test/unit/controllers/PublicationController.test.ts:27]: the response can be defined describe level, save you repeating the same const every test

File: src/test/unit/controllers/PublicationController.test.ts (hunk 2 above)
Comment (kurt-uk) [inline src/test/unit/controllers/PublicationController.test.ts:78]: needs a better name that publications2

File: src/test/unit/controllers/PublicationController.test.ts [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+  it('should send data if only one pub is returned', async () => {
+    const response = {
+      send: function () {
+        return '';
+      },
+    } as unknown as Response;
+    const request = mockRequest(i18n);
+    request.query = {courtId: 'x'};
+    request.user = {id: 1};
+    sinon.reset();
```
Comment (kurt-uk) [inline src/test/unit/controllers/PublicationController.test.ts:88]: youve already got your reset higher up but it looks like you dont need yo reset your stub, you just need to assign it a var and then return different params when different args are passed
```
(class level) 
const publicationsStub = sinon.stub(PublicationSerivce.prototype, 'getPublications);
publicationsStub.withArgs(CONDTIONA).resolves(solutionA);
publicationStub.withArgs(CONDITIONB).resolves(solutionB);
```
then you wont have to keep resetting stubs

**Review by kurt-uk — CHANGES_REQUESTED** [review]

File: src/main/controllers/PublicationController.ts [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+  public async get(req: PipRequest, res: Response): Promise<void> {
+    const courtId = req.query['courtId'];
+    let courtName: string;
+    if (courtId === '0'){
+      courtName = 'Single Justice Procedure';
+    }else if(typeof courtId === 'string'){
+      courtName = courtId.toString();
+    }
+    else{
+      courtName = 'Court not found';
```
[the remark this answers is not in the document]
Comment (furnivall) [reply src/main/controllers/PublicationController.ts:21]: Hey man, yeah, to be honest, it's a temporary solution just to distinguish Single Justice Procedure (has a defined name) and other courts (are just a number).
In practice, it's the difference between this: 
image
and this:
image

[the remark this answers is not in the document]
Comment (furnivall) [reply src/main/controllers/PublicationController.ts]: After thinking about it a bit, I've rewritten this, so it now does an additional request to the courtService which solves things neatly (while also reducing the ugly ifs)

File: src/main/views/summary-of-publications.njk (hunk 5 above)
Comment (furnivall) [reply src/main/views/summary-of-publications.njk:32]: image
I was just building off the heroku implementation. Happy to do it with CSS instead if that's desired!

Comment (kurt-uk) [reply src/main/views/summary-of-publications.njk:32]: if possible yeh that would be neater

File: src/main/modules/nunjucks/index.ts [hunk 10]
```diff
@@ -37,6 +37,13 @@ export class Nunjucks {
       },
     );
 
+    const dateFilter = require('nunjucks-date-filter');
+    env.addFilter('date', dateFilter);
+    // const listTypeLookup = require('listTypeLookup');
```
Comment (kurt-uk) [inline src/main/modules/nunjucks/index.ts:42]: can remove?

File: src/main/resources/locales/en/account-home.json [hunk 11]
```diff
@@ -15,7 +15,7 @@
     {
       "title": "Single Justice Procedure cases",
       "description": "A nationwide list of cases, ready to be decided by a magistrate without a hearing. Such as TV licensing prosecutions.",
-      "link": "single-justice-procedure"
+      "link": "summary-of-publications?courtId=0"
```
Comment (kurt-uk) [inline src/main/resources/locales/en/account-home.json:18]: dont need in the language json, the link is the same in any language

File: src/main/resources/requests/utils/axiosConfig.ts [hunk 12]
```diff
@@ -1,4 +1,4 @@
 import axios from 'axios';
 
-export const dataManagementApi = axios.create({baseURL: (process.env.DATA_MANAGEMENT_URL || 'https://pip-data-management.staging.platform.hmcts.net/'), timeout: 10000});
+export const dataManagementApi = axios.create({baseURL: (process.env.DATA_MANAGEMENT_URL || 'http://localhost:8090'), timeout: 10000});
```
Comment (kurt-uk) [inline src/main/resources/requests/utils/axiosConfig.ts:3]: needs changing back

File: src/main/routes/routes.ts [hunk 13]
```diff
@@ -119,7 +118,7 @@ export default function(app: Application): void {
     (req, res) => {res.redirect('/subscription-management');});
 
   //TODO: To be deleted/modified post UAT with suitable solution
-  app.get('/list-option', app.locals.container.cradle.listOptionController.get);
+  app.get('/summary-of-publications', app.locals.container.cradle.summaryOfPublicationsController.get);
```
Comment (kurt-uk) [inline src/main/routes/routes.ts:121]: can be moved out of the unsuitable solution to the other journeys

**Review by kurt-uk — CHANGES_REQUESTED** [review]

File: src/test/unit/controllers/SummaryOfPublicationsController.test.ts [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+const CourtStub = sinon.stub(CourtService.prototype, 'getCourtById');
+
+describe('Get publications', () => {
+  CourtStub.withArgs(0).resolves(JSON.parse('{"name":"Single Justice Procedure (SJP)"}'));
+  CourtStub.withArgs(1).resolves(JSON.parse('{"name":"New Court"}'));
+  sinon.stub(PublicationService.prototype, 'getPublications').resolves(sjpCases);
+
+  afterAll(function () {
+    sinon.restore();
+  });
```
Comment (kurt-uk) [inline src/test/unit/controllers/SummaryOfPublicationsController.test.ts:27]: bin

**Review by ChrisS1512 — DISMISSED** [review]

**Review by ChrisS1512 — APPROVED** [review]

File: src/test/e2e/tests/specs.test.ts [hunk 15]
```diff
@@ -113,11 +107,7 @@ describe('Unverified user', () => {
 
       it('selecting first result should take you to to the hearings list page', async () => {
         hearingListPage = await alphabeticalSearchPage.selectFirstListResult();
-        expect(await hearingListPage.getPageTitle()).toEqual('Blackburn Magistrates\' Court hearing list');
-      });
-
-      it(`should display ${expectedNumOfHearings} results`, async() => {
-        expect(await hearingListPage.getResults()).toBe(expectedNumOfHearings);
+        expect(await hearingListPage.getPageTitle()).toEqual('Summary of Publications for Blackburn Magistrates\' Court');
```
Comment (kurt-uk) [inline src/test/e2e/tests/specs.test.ts:110]: youve still got the old flow in here, this isnt the hearing list page, theres a full new page added here that sits in the flow which means the test of the number of results should be moved further along after something is clicked?

File: src/test/e2e/tests/specs.test.ts [hunk 16]
```diff
@@ -156,7 +146,7 @@ describe('Unverified user', () => {
     it('should select \'Single Justice Procedure case\' option and navigate to Single Justice Procedure case page', async () => {
       await viewOptionPage.selectOption('SingleJusticeProcedureRadioButton');
       singleJusticeProcedurePage = await viewOptionPage.clickContinueSingleJusticeProcedure();
-      expect(await singleJusticeProcedurePage.getPageTitle()).toEqual('Single Justice Procedure cases');
+      expect(await singleJusticeProcedurePage.getPageTitle()).toEqual('Summary of Publications for Single Justice Procedure (SJP)');
```
Comment (kurt-uk) [inline src/test/e2e/tests/specs.test.ts:149]: same here, this is the step before the SJP page right? theres a test that sits before this one to get to the page that the test was origionally talking about

File: src/test/routes/single-justice-procedure.test.ts [hunk 17]
```diff
@@ -6,7 +6,7 @@ describe('Single Justice Procedure', () => {
   describe('on GET', () => {
     test('should return single-justice-procedure page', async () => {
       await request(app)
-        .get('/single-justice-procedure')
+        .get('/summary-of-publications?courtId=0')
```
Comment (kurt-uk) [inline src/test/routes/single-justice-procedure.test.ts:9]: is this a single justice procedure?? this is summary of publications pertaining to a SJP, not the SJP page right? this is just the same SOP page but for a specific court

File: src/test/unit/requests/publicationRequests.test.ts [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+import {dataManagementApi} from '../../../main/resources/requests/utils/axiosConfig';
+
+const rawDataPubs = fs.readFileSync(path.resolve(__dirname, '../../../test/unit/mocks/summaryOfPublications.json'), 'utf-8');
+const pubs = JSON.parse(rawDataPubs);
+const totalCases = 3;
+const pubRequests = new SummaryOfPublicationsRequests();
+const stub = sinon.stub(dataManagementApi, 'get');
+
+beforeEach(async () => {
+  stub.withArgs('/publication/search/0').resolves({data: pubs});
```
Comment (kurt-uk) [inline src/test/unit/requests/publicationRequests.test.ts:14]: you only need to do this stub definition once, youre doing it before each test at the mo, should be able to be moved with your stub definition and the before each deleted

**Review by kurt-uk — CHANGES_REQUESTED** [review]

File: src/test/unit/requests/publicationRequests.test.ts (hunk 18 above)
Comment (furnivall) [reply src/test/unit/requests/publicationRequests.test.ts:14]: Should be sorted now :)

**Review by kurt-uk — DISMISSED** [review]

**Review by ChrisS1512 — DISMISSED** [review]

**Review by alberto-hortiguela — DISMISSED** [review]

**Review by commitBlob — APPROVED** [review]

**Review by alberto-hortiguela — APPROVED** [review]

**Pull request merged**